### PR TITLE
server: pretty region log

### PR DESCRIPTION
**Pull request by nolouch**
close https://github.com/pingcap/pd/issues/546
PTAL @siddontang @disksing

### CODE REVIEW

File: server/cache.go [hunk 1]
```diff
@@ -526,12 +527,12 @@ func (c *clusterInfo) handleRegionHeartbeat(region *regionInfo) error {
...
 	// Region meta is updated, update kv and cache.
 	if r.GetVersion() > o.GetVersion() || r.GetConfVer() > o.GetConfVer() {
-		log.Infof("update region %v origin %v", region, origin)
+		log.Infof("[region %d] %s, Epoch changed from {%s} to {%s},", region.GetId(), strings.Join(origin.Diff(region), ","), o, r)
 		return c.putRegionLocked(region)
 	}
 
 	if region.Leader.GetId() != origin.Leader.GetId() {
-		log.Infof("update region leader %v origin %v", region, origin)
+		log.Infof("[region %d] Leader changed from peer:{id:%d store_id:%d} to peer:{id:%d store_id:%d}", region.GetId(), origin.Leader.GetId(), origin.Leader.GetStoreId(), region.Leader.GetId(), region.Leader.GetStoreId())
```
Comment (siddontang) [inline server/cache.go]: `from %v to %v`, print the protobuf Peer directly.

File: server/operator.go [hunk 2]
```diff
@@ -116,7 +116,7 @@ func (op *regionOperator) GetResourceKind() ResourceKind {
 
 func (op *regionOperator) Do(region *regionInfo) (*pdpb.RegionHeartbeatResponse, bool) {
 	if time.Since(op.Start) > maxOperatorWaitTime {
-		log.Errorf("%s : operator timeout", op)
+		log.Errorf("Operator timeout:%s", op)
```
Comment (siddontang) [inline server/operator.go]: add `[region %d]` here.

File: server/cache.go [hunk 3]
```diff
@@ -526,12 +527,12 @@ func (c *clusterInfo) handleRegionHeartbeat(region *regionInfo) error {
 
 	// Region meta is updated, update kv and cache.
 	if r.GetVersion() > o.GetVersion() || r.GetConfVer() > o.GetConfVer() {
-		log.Infof("update region %v origin %v", region, origin)
+		log.Infof("[region %d] %s, Epoch changed from {%s} to {%s},", region.GetId(), strings.Join(origin.Diff(region), ","), o, r)
```
Comment (siddontang) [inline server/cache.go]: log ConfVer and version changed separately.

File: server/cache.go [hunk 4]
```diff
@@ -525,13 +525,17 @@ func (c *clusterInfo) handleRegionHeartbeat(region *regionInfo) error {
 	}
 
 	// Region meta is updated, update kv and cache.
-	if r.GetVersion() > o.GetVersion() || r.GetConfVer() > o.GetConfVer() {
-		log.Infof("update region %v origin %v", region, origin)
+	if r.GetVersion() > o.GetVersion() {
+		log.Infof("[region %d] %s, Version changed from {%d} to {%d}", region.GetId(), origin.DiffKey(region), o.GetVersion(), r.GetVersion)
```
Comment (siddontang) [inline server/cache.go]: maybe r.GetVersion()

Comment (siddontang) [inline server/cache.go]: I think print range changed from [start, end) to [start, end) is enough, no need DiffKey here.

Comment (siddontang) [inline server/cache.go]: Maybe we can add a RangeStr function to return `[start, end)`.

Comment (siddontang) [inline server/cache.go]: You can `range %v -> %v, version %d -> %d`

File: server/region.go [hunk 5]
```diff
@@ -59,6 +60,50 @@ func (r *regionInfo) String() string {
 	return fmt.Sprintf("%+v", *r)
 }
 
+func (r *regionInfo) DiffPeer(other *regionInfo) string {
```
Comment (siddontang) [inline server/region.go]: I think for DiffPeers, you should return two []Peer for added and removed, not string here.

Btw, seem that it is complex here, so I prefer logging like `peers changed %v -> %v` with Peers directly.

File: server/cache.go [hunk 6]
```diff
@@ -526,11 +526,11 @@ func (c *clusterInfo) handleRegionHeartbeat(region *regionInfo) error {
 
 	// Region meta is updated, update kv and cache.
 	if r.GetVersion() > o.GetVersion() {
-		log.Infof("[region %d] %s, Version changed from {%d} to {%d}", region.GetId(), origin.DiffKey(region), o.GetVersion(), r.GetVersion)
+		log.Infof("[region %d] %s, Version changed from {%d} to {%d}", region.GetId(), diffRegionPeersInfo(origin, region), o.GetVersion(), r.GetVersion())
 		return c.putRegionLocked(region)
 	}
 	if r.GetConfVer() > o.GetConfVer() {
-		log.Infof("[region %d] %s, ConfVer changed from {%d} to {%d}", region.GetId(), origin.DiffPeer(region), o.GetConfVer(), r.GetConfVer)
+		log.Infof("[region %d] %s, ConfVer changed from {%d} to {%d}", region.GetId(), diffRegionPeersInfo(origin, region), o.GetConfVer(), r.GetConfVer())
```
Comment (siddontang) [inline server/cache.go]: diffRegionKeyInfo

File: server/cache.go [hunk 7]
```diff
@@ -525,13 +525,17 @@ func (c *clusterInfo) handleRegionHeartbeat(region *regionInfo) error {
...
 
 	// Region meta is updated, update kv and cache.
-	if r.GetVersion() > o.GetVersion() || r.GetConfVer() > o.GetConfVer() {
-		log.Infof("update region %v origin %v", region, origin)
+	if r.GetVersion() > o.GetVersion() {
+		log.Infof("[region %d] %s, Version changed from {%d} to {%d}", region.GetId(), diffRegionPeersInfo(origin, region), o.GetVersion(), r.GetVersion())
+		return c.putRegionLocked(region)
+	}
+	if r.GetConfVer() > o.GetConfVer() {
+		log.Infof("[region %d] %s, ConfVer changed from {%d} to {%d}", region.GetId(), diffRegionPeersInfo(origin, region), o.GetConfVer(), r.GetConfVer())
```
Comment (siddontang) [inline server/cache.go]: diffRegionKeyInfo

File: server/util.go [hunk 8]
```diff
@@ -371,3 +375,47 @@ func minDuration(a, b time.Duration) time.Duration {
 	}
 	return b
 }
+
+func diffRegionPeersInfo(origin *regionInfo, other *regionInfo) string {
```
Comment (siddontang) [inline server/util.go]: any test to cover it?

Comment (nolouch) [inline server/util.go]: yes

File: server/util.go [hunk 9]
```diff
@@ -371,3 +375,47 @@ func minDuration(a, b time.Duration) time.Duration {
...
+	}
+	return strings.Join(ret, ",")
+}
+
+func diffRegionKeyInfo(origin *regionInfo, other *regionInfo) string {
+	var ret []string
+	if !bytes.Equal(origin.Region.StartKey, other.Region.StartKey) {
+		originKey := &metapb.Region{StartKey: origin.Region.StartKey}
+		otherKey := &metapb.Region{StartKey: other.Region.StartKey}
+		ret = append(ret, fmt.Sprintf("StartKey Changed:%s -> %s", originKey, otherKey))
```
Comment (siddontang) [inline server/util.go]: why using metapb.Region to print the key?

Comment (nolouch) [inline server/util.go]: To print the protobuf escaped format

Comment (siddontang) [inline server/util.go]: can we use escape directly? I guess this print will output `Region`

Comment (siddontang) [inline server/util.go]: seem redundant

Comment (nolouch) [inline server/util.go]: it just prints start_key field without `Region`. protoc just expose Marshaler methods for Message. I need to do that.

**Comment by siddontang** [comment]
PTAL @disksing @overvenus

**Pull request merged**